# Curso de Introducción a la Ciencia de Datos

**Ejercicio SF Salaries**
Utilizaremos el `SF Salaries Dataset` Kaggle. Simplemente siga y complete las tareas que se detallan a continuación en negrita. Las tareas serán cada vez más difíciles a medida que avance.

La idea principal es aplicar pandas para organizar informacion, manipularla, buscar información en el dataset


**Importe pandas como pd. e imprima la version que esta usando**

In [2]:
# Copie el código aca

import pandas as pd

In [3]:
version_pd = pd.__version__
print("la versión de pandas es:", version_pd)

la versión de pandas es: 3.0.2


**leer el archivo Salaries.csv en un dataframe llamado salarios**

se encuentra en esta url y se puede leer directamente con pandas

https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv


In [4]:
# Copie el código aca

url = "https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv"
df = pd.read_csv(url)
df.head()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Agency,Status
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,NaN,San Francisco,NaN
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,NaN,San Francisco,NaN
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,NaN,San Francisco,NaN
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,NaN,San Francisco,NaN
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,NaN,San Francisco,NaN


**Use el método `.info()` para encontrar cuantas Columnas de atributos hay y que tipo de datos tiene cada una.**

In [5]:
# Copie el código aca
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  str    
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), str(3)
memory usage: 14.7 MB


- Segun la información del dataframe su tamaño es de *memory usage: 14.7+ MB*

- las columnas **Notes** y **Status** solamente tienen valores *NaN*, verifique que dice **0 non-null** osea 0 valores son diferentes a *NaN*
entonces todos los valores son *NaN* en esas columnas

**Elimine las columnas *Notes* y *Status* Recuerde usar el método de forma permanente (inplace), y luego visualice cuantas filas y columnas tiene luego de eliminar las columnas**

In [6]:
# Copie el código aca

df.drop(["Notes","Status"], axis=1, inplace=True)


**¿Cual es la nueva información del dataframe?**

In [7]:
df.shape
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


- Se puede verificar que al borrar las columnas que no tenian solo *NaN* se disminuye el tamaño del dataframe a: *memory usage: 12.5+ MB*

**Observe los primeros 5 registros del DataFrame.**

In [8]:
# Copie el código aca
df.head()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,San Francisco
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,San Francisco
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,San Francisco
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,San Francisco
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,San Francisco


**Observe los ultimos 5 registros del DataFrame.**

In [9]:
# Copie el código aca
df.tail()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
148649,148650,Roy I Tillery,Custodian,0.0,0.0,0.00,0.0,0.00,0.00,2014,San Francisco
148650,148651,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148651,148652,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148652,148653,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148653,148654,Joe Lopez,"Counselor, Log Cabin Ranch",0.0,0.0,-618.13,0.0,-618.13,-618.13,2014,San Francisco


**¿Cuantas filas y columnas tiene el dataframe?**

In [10]:
# Copie el código aca
print("El DataFrame tiene", df.shape[0], "filas y", df.shape[1], "columnas")

El DataFrame tiene 148654 filas y 11 columnas


**Observe los 10 registros que estan desde el indice 3000**

In [11]:
# Copie el código aca
df.loc[3000:3010]

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
3000,3001,WILLIAM BOYLE,FIREFIGHTER,105934.65,18530.98,16389.60,NaN,140855.23,140855.23,2011,San Francisco
3001,3002,ALAN LIN,ELECTRONIC MAINTENANCE TECHNICIAN,98965.80,41504.40,380.00,NaN,140850.20,140850.20,2011,San Francisco
3002,3003,WILLIAM CONLEY,SERGEANT III (POLICE DEPARTMENT),130457.78,3131.42,7256.39,NaN,140845.59,140845.59,2011,San Francisco
3003,3004,ROBERT MCDONALD,"LIEUTENANT, FIRE DEPARTMENT",123104.99,1833.56,15897.68,NaN,140836.23,140836.23,2011,San Francisco
3004,3005,MICHAEL COSTELLO,POLICE OFFICER III,112401.08,7957.74,20466.93,NaN,140825.75,140825.75,2011,San Francisco
3005,3006,EDNA HO,MANAGER V,140821.37,0.00,0.00,NaN,140821.37,140821.37,2011,San Francisco
3006,3007,RODRIGO CONEL,REGISTERED NURSE,126120.00,3981.78,10714.21,NaN,140815.99,140815.99,2011,San Francisco
3007,3008,VICTOR HURTADO,FIREFIGHTER,105934.67,13112.76,21763.50,NaN,140810.93,140810.93,2011,San Francisco
3008,3009,DONALD LANDRY,TRANSIT POWER LINE WORKER,92861.50,39789.32,8135.86,NaN,140786.68,140786.68,2011,San Francisco
3009,3010,FRANK LEE,"INSPECTOR III, (POLICE DEPARTMENT)",130481.56,725.40,9579.58,NaN,140786.54,140786.54,2011,San Francisco



La columna *Benefits* tiene valores `NaN`, ¿Cuantos NaN tiene la columna *Benefits*? (TIP: Seleccione la Columna y vefique cuantos Na tiene `.isna()` luego sumelos `.sum()` )

In [12]:
# Copie el código aca
print("La columna 'Benefits' tiene", df[["Benefits"]].isna().sum())

La columna 'Benefits' tiene Benefits    36163
dtype: int64


Al parecer los valores faltantes en la columna Benefits es por que no se le pagan beneficios entonces:

**Reemplace los valores *NaN* de la Columna *Benefits* por el valor de cero (Recuerde usar el método de forma inplace) y mostrar los primeros 10 elementos de la lista**

In [13]:
# Copie el código aca
df["Benefits"] = df["Benefits"].fillna(0)
print(df["Benefits"].loc[148645:148653])

148645    0.0
148646    0.0
148647    0.0
148648    0.0
148649    0.0
148650    0.0
148651    0.0
148652    0.0
148653    0.0
Name: Benefits, dtype: float64


**¿Cual es la nueva información del dataframe?**

In [14]:
# Copie el código aca
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          148654 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


**Cuantos valores unicos hay en las columnas categoricas?**

In [49]:
# Copie el código aca

valores_unicos = df.select_dtypes(include=["object","category", "string"]).nunique()
print("los valores unicos en las columnas categoricas son:\n", valores_unicos)


los valores unicos en las columnas categoricas son:
 EmployeeName    110811
JobTitle          2159
Agency               1
dtype: int64


- la columna **Agency** solamente tiene un valor unico, entonces no aporta informacion, por lo tanto se puede eliminar

**Elimine la columna *Agency* del dataframe**

In [52]:
# Copie el código aca
df.drop(columns = ["Agency"], inplace=True)

In [55]:
# Revisar el resultado de la eliminación de la columna "Agency"
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          148654 non-null  int64  
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
dtypes: float64(5), int64(3), str(2)
memory usage: 11.3 MB


**Cual es el promedio de BasePay ?**

In [68]:
# Copie el código aca
promedio_basepay = df["BasePay"].mean()
print(f"El promedio de BasePay es: {promedio_basepay:.2f}")

El promedio de BasePay es: 66325.45


**Cual es la cantidad mas alta de OvertimePay en el dataset ?**

In [69]:
# Copie el código aca
cantidad_overtime = df["OvertimePay"].max()
print(f"La cantidad máxima de OvertimePay es: {cantidad_overtime:.2f}")

La cantidad máxima de OvertimePay es: 245131.88


**Realice una descripcion estadistica de las columnas numericas?**

In [96]:
# Copie el código aca
columnas_numericas = df.select_dtypes(include=['float64'])
df[columnas_numericas.columns].describe()

,BasePay,OvertimePay,OtherPay,TotalPay,TotalPayBenefits
count,148045.000000,148650.000000,148650.000000,148654.000000,148654.000000
mean,66325.448840,5066.059886,3648.767297,74768.321972,93692.554811
std,42764.635495,11454.380559,8056.601866,50517.005274,62793.533483
min,-166.010000,-0.010000,-7058.590000,-618.130000,-618.130000
25%,33588.200000,0.000000,0.000000,36168.995000,44065.650000
50%,65007.450000,0.000000,811.270000,71426.610000,92404.090000
75%,94691.050000,4658.175000,4236.065000,105839.135000,132876.450000
max,319275.010000,245131.880000,400184.250000,567595.430000,567595.430000


**En este dataset las variables `Id` y `Year` representan variables categorica, convertir estas columnas a su formato correcto**

In [78]:
# Copie el código aca
df[["Id", "Year"]] = df[["Id", "Year"]].astype("category")


**la variable `Year` es categorica ordinal, primero detecte cuales son los valores unicos que tiene la columna**

In [84]:
# Copie el código aca
valores_uni_year = df["Year"].unique()
print("Valores únicos en la columna 'Year':\n", valores_uni_year)


Valores únicos en la columna 'Year':
 [2011, 2012, 2013, 2014]
Categories (4, int64): [2011, 2012, 2013, 2014]


**Apartir de la información anterior convierta la columna `Year` en variable categorica ordinal y visualice la columna**

In [91]:
# Copie el código aca
# En el codigo anterior se convirtieron las columnas "Id" y "Year" a tipo categorico, de igual manera se aplica el codigo
# nuevamente para verificar que se hayan convertido correctamente la columna "Year":

df["Year"] = df["Year"].astype("category")


In [94]:
print("La columna 'Year' tiene el tipo de dato:", df["Year"].dtypes)


La columna 'Year' tiene el tipo de dato: category


**Realice nuevamente una descripcion estadistica de las columnas numericas?**

In [97]:
# Copie el código aca
columnas_numericas = df.select_dtypes(include=['float64'])
df[columnas_numericas.columns].describe()

,BasePay,OvertimePay,OtherPay,TotalPay,TotalPayBenefits
count,148045.000000,148650.000000,148650.000000,148654.000000,148654.000000
mean,66325.448840,5066.059886,3648.767297,74768.321972,93692.554811
std,42764.635495,11454.380559,8056.601866,50517.005274,62793.533483
min,-166.010000,-0.010000,-7058.590000,-618.130000,-618.130000
25%,33588.200000,0.000000,0.000000,36168.995000,44065.650000
50%,65007.450000,0.000000,811.270000,71426.610000,92404.090000
75%,94691.050000,4658.175000,4236.065000,105839.135000,132876.450000
max,319275.010000,245131.880000,400184.250000,567595.430000,567595.430000


**¿Cuál es el título de trabajo de JOSEPH DRISCOLL? Nota: Utilice mayúsculas, de lo contrario, puede obtener una respuesta que no coincida.**

In [101]:
# Copie el código aca
print("EL titulo de trabajo de JOSEPH DRISCOLL es:", df[df["EmployeeName"] == "JOSEPH DRISCOLL"]["JobTitle"].values[0])

EL titulo de trabajo de JOSEPH DRISCOLL es: CAPTAIN, FIRE SUPPRESSION


**¿Cuanto gana JOSEPH DRISCOLL (incluyendo beneficios)?**

In [111]:
# Copie el código aca
print("El salario total incluyendo beneficios de JOSEPH DRISCOLL es:", df[df["EmployeeName"] == "JOSEPH DRISCOLL"]["TotalPayBenefits"].values[0], "USD")

El salario total incluyendo beneficios de JOSEPH DRISCOLL es: 270324.91 USD


**¿Cual es el nombre de la persona con mejor pago (incluyendo beneficios)?**

In [121]:
# Copie el código aca
display("El empleado con el salario total más alto es:\n", df.loc[df["TotalPayBenefits"].idxmax()])

'El empleado con el salario total más alto es:\n'

Id                                                               1
EmployeeName                                        NATHANIEL FORD
JobTitle            GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY
BasePay                                                  167411.18
OvertimePay                                                    0.0
OtherPay                                                 400184.25
Benefits                                                         0
TotalPay                                                 567595.43
TotalPayBenefits                                         567595.43
Year                                                          2011
Name: 0, dtype: object

**¿Cual es el nombre de la persona peor paga (incluyendo beneficios)? ¿Notan algo extraño sobre cuánto le pagan?**

In [120]:
# Copie el código aca
display("El empleado con el salario total más bajo es:\n", df.loc[df["TotalPayBenefits"].idxmin()])

'El empleado con el salario total más bajo es:\n'

Id                                      148654
EmployeeName                         Joe Lopez
JobTitle            Counselor, Log Cabin Ranch
BasePay                                    0.0
OvertimePay                                0.0
OtherPay                               -618.13
Benefits                                     0
TotalPay                               -618.13
TotalPayBenefits                       -618.13
Year                                      2014
Name: 148653, dtype: object

**Cuantos registros tienen un valor de TotalPayBeneficts menor que cero**

In [124]:
# Copie el código aca
registros_neg = df[df["TotalPayBenefits"] < 0]
cantidad_neg = registros_neg.shape[0]
print(f"La cantidad de registros con TotalPayBenefits negativo es: {cantidad_neg}")
display(registros_neg)

La cantidad de registros con TotalPayBenefits negativo es: 4


,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year
110528,110529,Timothy E Gibson,Police Officer 3,NaN,0.0,0.00,-2,0.00,-2.73,2013
110529,110530,Mark E Laherty,Police Officer 3,NaN,0.0,0.00,-8,0.00,-8.20,2013
110530,110531,David P Kucia,Police Officer 3,NaN,0.0,0.00,-33,0.00,-33.89,2013
148653,148654,Joe Lopez,"Counselor, Log Cabin Ranch",0.0,0.0,-618.13,0,-618.13,-618.13,2014


**¿Cual es el promedio (mean) BasePay de todos los empleados por año? los años van entre (2011-2014)?** (Tip: Usar groupby)

In [127]:
# Copie el código aca
print("El salario promedio por año es:\n", df.groupby("Year")["TotalPayBenefits"].mean())

El salario promedio por año es:
 Year
2011     71744.103871
2012    100553.229232
2013    101440.519714
2014    100250.918884
Name: TotalPayBenefits, dtype: float64


**¿Cuantos titulos de trabajos (JobTitle) unicos existen?**

In [ ]:
# Copie el código aca
profesiones_unicas = df["JobTitle"].nunique()
print("La cantidad de profesiones únicas es:", profesiones_unicas)

La cantidad de profesiones únicas es: 2159


**¿Cuales son los 5 trabajos mas comunes?**

In [137]:
# Copie el código aca
profesiones_comunes = df["JobTitle"].value_counts().head(5)
display("Las 5 profesiones más comunes son:\n", profesiones_comunes)

'Las 5 profesiones más comunes son:\n'

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64

**¿Cuántos puestos de trabajo Unicos existieron en 2013? (Ejemplo, los títulos de trabajo con solo una ocurrencia en 2013?)**

In [141]:
# Copie el código aca
profesiones_comunes_2013 = df[df["Year"] == 2013]["JobTitle"].value_counts().head()
display("Las profesiones más comunes en el año 2013 son:\n", profesiones_comunes_2013)

'Las profesiones más comunes en el año 2013 son:\n'

JobTitle
Transit Operator                2295
Special Nurse                   1445
Registered Nurse                1231
Public Svc Aide-Public Works     859
Custodian                        789
Name: count, dtype: int64

**¿Cuántas personas tienen la palabra Chief en el título de su trabajo?** Apoyese en los métodos para manipular strings y en el método `.apply()`

In [142]:
# Copie el código aca
personas_chief = df[df["JobTitle"].apply(lambda x: "chief" in x.lower())]
cantidad_chief = personas_chief.shape[0]
print(f"La cantidad de personas con 'chief' en su título de trabajo es: {cantidad_chief}")

La cantidad de personas con 'chief' en su título de trabajo es: 627


Este código es valido, se recomienda hacer el código mas simple y agregar typing y documentación.

In [144]:
def contar_chief(job_title):
    return "chief" in job_title.lower()
print("La cantidad de personas con 'chief' en su título de trabajo es:", df["JobTitle"].apply(contar_chief).sum())

La cantidad de personas con 'chief' en su título de trabajo es: 627


**Bono: ¿Existe una correlación entre la longitud del nombre del Título del trabajo y el Salario?**

crea una nueva columna donde se guarde la longitud del titulo del trabajo. tip: usar el método `.apply()`

In [151]:
# Copie el código aca
df["Long_JobTitle"] = df["JobTitle"].apply(len)
display(df[["JobTitle" , "Long_JobTitle"]].head())

,JobTitle,Long_JobTitle
0,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,46
1,CAPTAIN III (POLICE DEPARTMENT),31
2,CAPTAIN III (POLICE DEPARTMENT),31
3,WIRE ROPE CABLE MAINTENANCE MECHANIC,36
4,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",44


Utilice el método `.corr()` de Pandas para calcular la correlacion

In [155]:
# Copie el código aca
correlacion = df[["Long_JobTitle", "TotalPayBenefits"]].corr()
display("La correlación entre la longitud del título de trabajo y el salario total es:\n", correlacion)

'La correlación entre la longitud del título de trabajo y el salario total es:\n'

,Long_JobTitle,TotalPayBenefits
Long_JobTitle,1.000000,-0.036878
TotalPayBenefits,-0.036878,1.000000
